File: java/gatherData/NaverKeywords.java
package org.gatherData;

import java.lang.Thread;
import java.time.Duration;
import java.time.LocalDateTime;
import org.json.JSONArray;
import org.json.JSONObject;
import org.config.SeleniumSettings;

public class NaverKeywords {
    String   naverSearchUrl;
    String[] keywordsList;

    public JSONArray        naverDataStatusArray, naverDataDictionary;
    public SeleniumSettings seleniumObject;

    NaverKeywords(SeleniumSettings seleniumObject) {
        this.seleniumObject = seleniumObject;
        seleniumObject.driverSettings();
    }

    public void naverKeywordsSettingsMethod(String naverSearchUrl, String[] keywordsList) {
        this.naverSearchUrl       = naverSearchUrl;
        this.keywordsList         = keywordsList;
        this.naverDataStatusArray = new JSONArray();
        this.naverDataDictionary  = new JSONArray();
    }

    int saveStatusData(String keyword, String[] extractedKeywords) {
        JSONObject singleInnerJson  = new JSONObject();
        Integer    extractionStatus = ((extractedKeywords.length >= 1) && (!extractedKeywords[0].equals("extracted_failed"))) ? 1 : 0;

        singleInnerJson.put("date",              String.valueOf(LocalDateTime.now()).substring(0, 10));
        singleInnerJson.put("search_keyword",    keyword);
        singleInnerJson.put("completion_status", extractionStatus);
        naverDataStatusArray.put(singleInnerJson);

        return extractionStatus;
    }

    void saveCrawledData(String keyword, String[] extractedKeywords) {
        int counterVar = 1;

        for (String subKeyword : extractedKeywords) {
            if (!subKeyword.equals("추가")) {
                JSONObject naverDataJObject = new JSONObject();
                naverDataJObject.put("date",              String.valueOf(LocalDateTime.now()).substring(0, 10));
                naverDataJObject.put("search_keyword",    keyword);
                naverDataJObject.put("autocomplete_term", subKeyword);
                naverDataJObject.put("term_ranking",      String.format("%02d", counterVar));
                naverDataDictionary.put(naverDataJObject);
                counterVar += 1;
            }
        }
    }

    public void getKeywordData() throws InterruptedException, Exception {
        seleniumObject.driver.get(naverSearchUrl);
        Thread.sleep(Duration.ofSeconds(4));

        for (String keyword : keywordsList) {
            seleniumObject.sendStringToElement("/html/body/div[2]/div[1]/div/div[3]/div[2]/div/form/fieldset/div/input", "xpath", keyword);
            Thread.sleep(Duration.ofSeconds(4));
            String[] extractedKeywords = seleniumObject.searchForElement("/html/body/div[2]/div[1]/div/div[3]/div[2]/div/form/fieldset/div/input", "xpath").getText().split("\n");
            int      extractionStatus  = saveStatusData(keyword, extractedKeywords);
            String[] finalizedKeywords = (extractionStatus == 1) ? extractedKeywords : new String[] {"extraction_failed"};
            saveCrawledData(keyword, finalizedKeywords);
            seleniumObject.searchForElement("/html/body/div[2]/div[1]/div/div[3]/div[2]/div/form/fieldset/div/input", "xpath").clear();
        }
    }
}

File: src/gather_data/NaverKeywords.py
import time

import datetime as dt 

from config.SeleniumSettings import SeleniumSettings 

class NaverKeywords:
    def __init__(self, selenium_object: SeleniumSettings) -> None: 
        self.selenium_object = selenium_object 
        self.selenium_object.driver_settings()

    def naver_keywords_settings_method(self, naver_search_url: str, keywords_list: list[str]) -> None: 
        self.naver_search_url = naver_search_url 
        self.keywords_list    = keywords_list 
        
        self.naver_data_status_check = {
            "date"               : [],
            "search_keyword"     : [],
            "completion_status"  : []
        }

        self.naver_data_dictionary = {
            "date"              : [],
            "search_keyword"    : [],
            "autocomplete_term" : [],
            "term_ranking"      : []
        }

    def get_keyword_data(self) -> None: 
        def save_status_data(keyword: str, extracted_keywords: list[str]) -> int:
            extraction_status = 1 if ((len(extracted_keywords) >= 1) and (extracted_keywords[0] != "extraction_failed")) else 0

            self.naver_data_status_check["date"].append(str(dt.datetime.now())[0:10])
            self.naver_data_status_check["search_keyword"].append(keyword)
            self.naver_data_status_check["completion_status"].append(extraction_status)

            return extraction_status

        def save_crawled_data(keyword: str, extracted_keywords: str) -> None:
            for (rank_num, word) in enumerate(extracted_keywords):
                self.naver_data_dictionary["date"].append(str(dt.datetime.now())[0:10])
                self.naver_data_dictionary["search_keyword"].append(keyword)
                self.naver_data_dictionary["autocomplete_term"].append(word)
                self.naver_data_dictionary["term_ranking"].append(str(rank_num + 1).zfill(2))
        
        self.selenium_object.driver.get(self.naver_search_url)
        time.sleep(4)

        for keyword in self.keywords_list: 
            self.selenium_object.send_string_to_element("/html/body/div[2]/div[1]/div/div[3]/div[2]/div/form/fieldset/div/input", keyword)  
            time.sleep(4)
            extracted_keywords = [i for i in self.selenium_object.search_for_element("/html/body/div[2]/div[1]/div/div[3]/div[2]/div/div[2]/div/div/div[2]/div[1]/ul").text.split("\n") if ("추가" not in i)]
            extraction_status  = save_status_data(keyword, extracted_keywords)
            save_crawled_data(keyword, (lambda extraction_status, extracted_keywords: extracted_keywords if (extraction_status == 1) else ["extraction_failed"])(extraction_status, extracted_keywords))
            self.selenium_object.search_for_element("/html/body/div[2]/div[1]/div/div[3]/div[2]/div/form/fieldset/div/input").clear()
